android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/managers/AndroidBasePackageManager.java

```diff
@@ -0,0 +1,578 @@
+            // check versions of package
+            if (apkVersionCode == installedVersionCode
+                    && apkPackageInfo.versionName.equals(installedPackageInfo.versionName)) {
+                logger.atDebug().log("Package {} with same version and versionCode is already installed",
+                        packageName);
+                if (!force) {
+                    updateAPKInstalled(packageName, true);
+                    return;
+                }
+                logger.atDebug().log("Force flag is set, reinstall package {}", packageName);
+            }
+        }
+
+        // check for interruption
+        if (Thread.currentThread().isInterrupted()) {
+            logger.atWarn().log("Refusing to install because the active thread is interrupted");
+            throw new InterruptedException();
+        }
+
+        boolean uninstalled = false;
+        // check is uninstall required
+        if (installedVersionCode != -1 && apkVersionCode < installedVersionCode) {
+            logger.atDebug().log("Uninstalling package {} first due to downgrade is required from {} to {}",
+                    packageName, installedVersionCode, apkVersionCode);
+            uninstallPackage(packageName, logger);
+            updateAPKInstalled(packageName, false);
+            uninstalled = true;
+        }
+
+        // check for interruption but only when package was not uninstalled
+        if (!uninstalled && Thread.currentThread().isInterrupted()) {
+            logger.atWarn().log("Refusing to install because the active thread is interrupted");
+            throw new InterruptedException();
+        }
+
+        // finally install APK without checks
+        installAPK(apkPath, packageName, lastUpdateTime, logger);
+    }
+
+    /**
+     * Install APK file.
+     *
+     * @param apkPath   path to APK file
+     * @param packageName APK should contains that package
+     * @param lastUpdateTime previous package last update time to check for installation
+     * @param logger optional logger to log to
+     * @throws IOException      on errors
+     * @throws InterruptedException when thread has been interrupted
+     */
+    // TODO: android: rework with PackageInstaller
+    private void installAPK(@NonNull String apkPath, @NonNull String packageName,
+                            long lastUpdateTime, Logger logger)
+            throws IOException, InterruptedException {
+        File apkFile = new File(apkPath);
+        Intent intent = new Intent(ACTION_VIEW);
+        Context context = contextProvider.getContext();
+        Uri downloadedApk = FileProvider.getUriForFile(
+                context,
+                context.getPackageName() + PROVIDER,
+                apkFile);
+        intent.setDataAndType(downloadedApk, PACKAGE_ARCHIVE);
+        intent.setFlags(FLAG_ACTIVITY_NEW_TASK | FLAG_GRANT_READ_URI_PERMISSION);
+        context.startActivity(intent);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

When launching an Android activity with `startActivity(Intent, ...)`, confirm that an application on the device can receive the implicit intent and launch the activity. If an application cannot, then your application crashes when `startActivity` is called. See the following diffs on GitHub where this issue was fixed: [diff1](https://github.com/jtribe/shelly/commit/a51e6db04fe01da362d168465979461d0893529f#diff-6a81c9af078ea62d2a817e1b8de61a4f190d90530923d26d712eebf6c14dba62L113), [diff2](https://github.com/didi/DoraemonKit/commit/b6815f63159a564d3110242d78611fb61bdbf9b4#diff-86123abc1b9085ef0d59807e11f3aa9c0a0f0d138f7543dce034014c5276aa9fL89)

Learn more: [Common Intents in the Android developer guide](https://developer.android.com/guide/components/intents-common)